packages/protocol/contracts/libs/LibDeploy.sol

```diff
@@ -17,8 +17,19 @@ pragma solidity 0.8.24;
 import "lib/openzeppelin-contracts/contracts/proxy/ERC1967/ERC1967Proxy.sol";
 import "lib/openzeppelin-contracts-upgradeable/contracts/access/OwnableUpgradeable.sol";
 
+/// @title TaikoERC1967Proxy
+/// @dev A proxy that allows for reading its implementation address.
+contract TaikoERC1967Proxy is ERC1967Proxy {
+    constructor(address _logic, bytes memory _data) payable ERC1967Proxy(_logic, _data) { }
+
+    function implementation() public view returns (address) {
```

Comment:
If we explosed this address, then we can merge [this breaking change](https://github.com/taikoxyz/taiko-mono/pull/15609) without modifying the relayer's mk-proof data and only need to modify the Bridge dapp to check the bridge implementation address, if the impl is the old one, we use the bridge's return data as-is, otherwise, RLP-decode the date first into a bytes[] then use the data.

Reply:
I still don't really follow, shouldn't the proof data be versioned then? Because the smart contract gets some data from offchain in some format, and so may have to convert that data or not to work with the new Merkle trie lib. The current implementation used onchain doesn't say anything about the proof format supplied from offchain. Unless we of course simply force the bridge to use the updated format and so then it will have to do the conversion offchain by reading the implementation contract with `eth_getstorageat` and so supplying the expected format onchain?

Reply:
There are multiple approaches possible, one of them is the dapp queries the underlying SignalService implemenation address and if it is address(abc), then it queries the relayer with an extra data "?ss_addr=abc". The relayer then convert the data to match the address abc.

Reply:
But this is completely offchain no? No need to expose the implementation contract in a way that works onchain?

Reply:
Having this new function doesn't hurt, right? And it may be very useful in some cases. So why not have it?

Reply:
Right, but `eth_getstorageat` can get you the same value without any changes?

I don't have anything against exposing the implementation contract onchain btw! I just don't think it's required to do what needs to be done.

Reply:
> Having this new function doesn't hurt, right? And it may be very useful in some cases. So why not have it?

Yeah agree, the deployment cost for the proxy is not important for us.

Reply:
If we have this `implementation()` function now on A6, then we can modify the Bridge dapp as follows:

- query implementation(), it returns either address_a (old) or address_b (new with the new OP code) for SignalService on, for example L1. The dapp doesn't know exactly when the upgrade happends, it doesn't have to know.
- the dapp get the proof from the relayer.
- in case of address_a, do nothing, just use whatever the relayer returns
- in case of address_b, inside the dapp, convert the relayer's proof_data into the new format, then use it.

Reply:
Ok, maybe it's not right :(

Reply:
Yeah I got the idea, alternatively instead of calling `implementation()` the bridge dapp could also just call `eth_getstorageat`, but that is indeed more exotic so having the view function there may be simpler so seems good.